scipy/stats/tests/test_sobol_.py

```diff
@@ -0,0 +1,219 @@
+from collections import Counter
+
+import numpy as np
+import pytest
+
+from scipy.stats.qmc import Sobol
+
+from numpy.testing import (assert_, assert_equal, assert_array_almost_equal,
+                           assert_array_equal)
+from pytest import raises as assert_raises
+import scipy.stats.qmc as qmc
+
+
+class TestSobol:
+    # set maxDiff to None to show all differences when tests fail
+    maxDiff = None
+
+    def setUp(self):
+        with pytest.warns(UserWarning):
+            engine_unscrambled_1d = Sobol(1, scramble=False)
+            self.draws_unscrambled_1d = engine_unscrambled_1d.random(10)
+            engine_unscrambled_3d = Sobol(3, scramble=False)
+            self.draws_unscrambled_3d = engine_unscrambled_3d.random(10)
+            engine_scrambled_1d = Sobol(1, scramble=True, seed=12345)
+            self.draws_scrambled_1d = engine_scrambled_1d.random(10)
+            engine_scrambled_3d = Sobol(3, scramble=True, seed=12345)
+            self.draws_scrambled_3d = engine_scrambled_3d.random(10)
+
+    def test_warning(self):
+        with pytest.warns(UserWarning):
+            engine = Sobol(1)
+            engine.random(10)
+
+    def test_random_base2(self):
+        self.setUp()
+        engine = Sobol(1, scramble=False)
+        sample = engine.random_base2(2)
+        assert_array_equal(self.draws_unscrambled_1d[:4],
+                           sample)
+
+        # resampling still having N=2**n
+        sample = engine.random_base2(2)
+        assert_array_equal(self.draws_unscrambled_1d[4:8],
+                           sample)
+
+        # resampling again but leading to N!=2**n
+        assert_raises(ValueError, engine.random_base2, 2)
```

Comment:
Exactly.

Yes it is ok to use the second sample of size 4 as you skipped 4 point. But having just skipped 4, you can only use the next 4 points. You could also have used the first 2 samples of the second 4. Hope I am clear 😅

Reply:
Yes, I understand.  You have been very careful in the documentation and in warnings to explain that all points in the sequence must be used together.
On the other hand, the `fast_forward` method is provided. This sends a bit of a mixed message. 

Considering it all together, I still think many users would still make mistakes calling `random` multiple times to get what they think are independent samples, just as they do with instances of `RandomState` and `Generator`. Is there an example, other than a limited memory application, in which it is particularly important or convenient to generate a few samples at a time rather than generating them all at once (and perhaps using only portions at a time)?  If users are far more likely to make mistakes with the ability to call `random` multiple times than they are to use this ability to their advantage, it might make sense not to allow it.

Reply:
`fast_forward` is provided for consistency. It is not wrong to use this method if you know what you are doing: like skipping 2**M points allowing you to take at most 2**m, with m < M.

We came up with this `random_base2` to help the user.

If we want the user to be able to make repeated calls to the generator to have independent samples, we could reset by default the state all the time. But if we do this, you will for sure have most user mixing the samples from separate draw and there would be no warning doing so.

Sobol' is very tricky and if you want to benefit from it, you must follow some rules. With the current API, the user will have some warnings along the way when trying to do something else. I did not see any other implementation checking for all these rules and in the end we see users using Sobol' like any other generator.

Reply:
`fast_forward` is provided for consistency. It is not wrong to use this method if you know what you are doing: like skipping 2^M points allowing you to take at most 2^m, with m < M.

We came up with this `random_base2` to help the user.

If we want the user to be able to make repeated calls to the generator to have independent samples, we could reset by default the state all the time. But if we do this, you will for sure have most user mixing the samples from separate draw and there would be no warning doing so.

Sobol' is very tricky and if you want to benefit from it, you must follow some rules. With the current API, the user will have some warnings along the way when trying to do something else. I did not see any other implementation checking for all these rules and in the end we see users using Sobol' like any other generator.

Reply:
> you will for sure have most user mixing the samples from separate draw and there would be no warning doing so.

Ok. I'm just not familiar with that sort of scenario. Whenever I need random data, even from `np.random`, I generate everything I need in one call. Typically I use it all at once, too. I just can't think of an application in which I've mixed data from separate calls in a way that would be inappropriate. 

Go ahead and leave it as-is; I'll just want to see what another maintainer thinks before we merge.